src/mono/wasm/build/WasmApp.targets

```diff
@@ -1,31 +1,84 @@
+      <RuntimeEmccVersionFile>$(MicrosoftNetCoreAppRuntimePackRidDir)native\src\emcc-version.txt</RuntimeEmccVersionFile>
+      <EmccPublishVersionFile>$(WasmAOTDir)emcc-version.txt</EmccPublishVersionFile>
+    </PropertyGroup>
+    <ItemGroup>
+      <MonoAOTCompilerDefaultAotArguments Include="no-opt" />
+      <MonoAOTCompilerDefaultAotArguments Include="static" />
+      <MonoAOTCompilerDefaultAotArguments Include="direct-icalls" />
+      <MonoAOTCompilerDefaultAotArguments Include="deterministic" />
+    </ItemGroup>
+    <ItemGroup>
+      <AotInputAssemblies Include="@(_WasmAssemblies->Distinct())">
+        <AotArguments>@(MonoAOTCompilerDefaultAotArguments, ';')</AotArguments>
+        <ProcessArguments>@(MonoAOTCompilerDefaultProcessArguments, ';')</ProcessArguments>
+      </AotInputAssemblies>
+
+      <_WasmAssemblies Remove="@(_WasmAssemblies)" />
+    </ItemGroup>
+
+    <MonoAOTCompiler
+      CompilerBinaryPath="$(MicrosoftNetCoreAppRuntimePackRidDir)native\cross\mono-aot-cross"
+      Mode="LLVMOnly"
+      OutputType="AsmOnly"
+      Assemblies="@(AotInputAssemblies)"
+      UseAotDataFile="false"
+      AotModulesTablePath="$(WasmAOTDir)driver-gen.c"
+      UseLLVM="true"
+        DisableParallelAot="true"
+      LLVMPath="$(EMSDK_PATH)\upstream\bin">
+      <Output TaskParameter="CompiledAssemblies" ItemName="_WasmAssemblies" />
+    </MonoAOTCompiler>
+    <Message Text="_WasmAssemblies: @(_WasmAssemblies)" />
+  </Target>
+
+  <Target Name="_WasmResolveReferences" Condition="'$(WasmResolveAssembliesBeforeBuild)' == 'true'">
     <WasmLoadAssembliesAndReferences
-      Condition="'$(WasmResolveAssembliesBeforeBuild)' == 'true'"
-      Assemblies="$(WasmMainAssemblyPath);@(WasmAssembliesToBundle)"
+      Assemblies="@(_WasmAssemblies)"
       AssemblySearchPaths="@(WasmAssemblySearchPaths)"
       SkipMissingAssemblies="$(WasmSkipMissingAssemblies)">
-      <Output TaskParameter="ReferencedAssemblies" ItemName="_WasmAssembliesFullSet" />
+      <Output TaskParameter="ReferencedAssemblies" ItemName="_TmpWasmAssemblies" />
     </WasmLoadAssembliesAndReferences>
 
     <ItemGroup>
-      <_WasmAssembliesFullSet Condition="'$(WasmResolveAssembliesBeforeBuild)' != 'true'" Include="@(WasmAssembliesToBundle)" />
+      <_WasmAssemblies Remove="@(_WasmAssemblies)" />
+      <_WasmAssemblies Include="@(_TmpWasmAssemblies)" />
     </ItemGroup>
+  </Target>
+
+  <Target Name="_BeforeWasmBuildApp">
+      <Error Condition="'$(WasmMainAssemblyPath)' == ''" Text="%24(WasmMainAssemblyPath) property needs to be set" />
+      <Error Condition="'$(WasmAppDir)' == ''" Text="%24(WasmAppDir) property needs to be set" />
+      <Error Condition="'$(RunAOTCompilation)' == 'true' and '$(WasmAOTDir)' == ''" Text="%24(WasmAOTDir) property needs to be set" />
+      <Error Condition="!Exists('$(MicrosoftNetCoreAppRuntimePackRidDir)')" Text="MicrosoftNetCoreAppRuntimePackRidDir=$(MicrosoftNetCoreAppRuntimePackRidDir) doesn't exist" />
```

Comment:
Hm I wonder if we still need to set this explicitly in the projects. We could resolve all the other paths that currently use it to use `%(RuntimePack.PackageDirectory)` instead. That's also basically what we'll need to do for "real" usage of this target outside of dotnet/runtime.